Improve locale detection.

bruvzg opened the pull request:
Instead of using a hard-coded supported locale list and expecting OS locale to match, the engine now attempts to decompose whatever garbage is returned by the `OS.get_locale()`, infer missing parts and select the best matching locale.

Exposes `compare_locales` and `standardize_locale` methods to the scripts.

So far it's tested on macOS only, all supported editor translations seem to be detected correctly.

bruvzg committed:
Improve locale detection.
Use separate language, script and country lists.
Add locale selection dialog and property hint.

akien-mga commented on doc/classes/TranslationServer.xml:
```diff
@@ -24,6 +24,14 @@
+			<argument index="0" name="locale_a" type="String" />
+			<argument index="1" name="locale_b" type="String" />
+			<description>
+				Compares two locales and return similarity score between [code]0[/code](no match) and [code]10[/code](full macth).
```
```suggestion
				Compares two locales and return similarity score between [code]0[/code](no match) and [code]10[/code](full match).
```

akien-mga commented:
Thanks for working on this, it's been a sore point and the hardcoded list we had was far from exhaustive.

But I wonder - is there really any merit to limit what can be valid locales? In practice we mostly need locales to match translation identifiers, so if someone happens to have a locale of say `de_RU` (which is not "valid", but there could be German speakers in Russia ending up with something like this), they should simply be matched to `de.po`/`de` in the CSV. Or `de_DE` as fallback if `de` is missing.

So the only country-specific information we'd need to hardcode IMO is just what's the "main" country code to fall back to if there's no translation matching the exact locale:
- 1. Check against exact user locale: `de_RU`
- 2. Check for language code only: `de`
- 3. Check for language code with "main" country: `de_DE`
- 4. Check for any match to language code + anything else: `de_<whatever>`, first match?

bruvzg commented:
Now it should work with all supported locales on Linux (tested with the list in `/usr/share/i18n/SUPPORTED` on Ubuntu 21.04), Windows (https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-lcid/70feba9f-294e-491e-b6eb-56532684c37f) and macOS.

Changes in this PR:
- Uses separate language, script and country lists (same lists are reused by font import options).
- Modified `standardize_locale` function: return locale as "lng_Script_CN_variant" (some parts can be missing), it's doing following actions:
  - replace "-" with  "_".
  - removes unsupported script and country codes (OS can add your current country code to the locale).
  - parses extra part of locale string after "@" to detect regional variant (e.g. `ca_ES@valencia` ) or script (e.g. `uz_UZ@cyrillic`) and remove the rest (currency and collation tags).
  - adds default values of the script and country, in some cases it might be ambiguous.
  - resolves country and language renames.
  - checks if language code is valid (and fallback to "en" if it's not).
- Added `compare_locales` function: returns value from `0` to `10`, where `10` is full match (after both strings passed to `standardize_locale`), `0` is no match (if language code is different the rest is not checked), values in between indicate number of matching elements (including language).
- Removes `is_locale_valid` and error messages, since `standardize_locale` should always return something valid.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
-	while (*ptr) {
-		if (*ptr == p_locale) {
-			return true;
+static const char *locale_varinats[][2] = {
```
Typo "varinats".

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
+	{ nullptr, nullptr }
+};
+
+// Additional script information to preferred scripts.
+// Language code, script code, default country, supported countries.
```
Is there a specific reference you used for those?

bruvzg replied:
Same references as language lists.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
+	{ "US", "United States of America" },
+	{ "UY", "Uruguay" },
+	{ "UZ", "Uzbekistan" },
+	{ "VA", "Holy See" },
```
Lol, I never knew this was the official name.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
+
+// Languages code, language name, supported country codes.
+// Reference:
+// - https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes
+// - https://lh.2xlibre.net/locales/
+// - https://iso639-3.sil.org/
+// - https://www.localeplanet.com/icu/index.html
```
Are all of these still used to make this list? If not, it would be good to narrow down to a single source so that we know how to sync potential changes.

bruvzg replied:
I have used multiple sources to make sure every language supported by the OS is in the list, but some links the might be excessive and have the same list, I'll re-check it.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
+Map<String, String> TranslationServer::script_map;
+Map<String, String> TranslationServer::locale_rename_map;
+Map<String, String> TranslationServer::country_name_map;
+Map<String, String> TranslationServer::varinat_map;
```
Typo.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
-	if (split == -1) {
-		split = p_locale.find("-");
+String TranslationServer::standardize_locale(const String &p_locale) const {
+	// Replaces '-' with '_' for macOS Sierra-style locales.
```
Is the reference to "Sierra" still valid?

bruvzg replied:
`Sierra` is not relevant, it's used by all macOS versions.

akien-mga commented on core/string/translation.cpp:
```diff
@@ -1004,124 +212,1503 @@ static _character_accent_pair _character_to_accented[] = {
+// - https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
+// - https://www.iso.org/obp/ui/#search/code/
+
+static const char *country_names[][2] = {
```
Maybe we could move these data tables to a dedicated file, like `core/string/locales.h`?

bruvzg replied:
Moved to `core/string/locales.h`.

akien-mga commented:
Looks good to me overall, though I'm not sure if we actually get useful information from a hardcoded list of valid countries for a given language.

That's obviously a decision ISO and various OSes followed so there might be a good reason, but nothing prevents me from using an OS in French while living in Denmark and thus ending up with a `fr_DK` locale which isn't standard.

Do we do anything specific with the valid country codes information?

bruvzg commented:
> Do we do anything specific with the valid country codes information?

It's used in two places:
- to ignore non-standard variants,
- and to generate output of `get_all_locales` used to add translations in the "Localization Editor" and locale specific overrides for the font import, but this part might be changed to the separate country + language code selector, which is likely more convenient than a single long list anyway.

Not sure if checking for the valid language and country code is necessary at all. Just using `locale_scripts` and renames to resolve ambiguous cases should be enough.

bruvzg commented:
I have removed all "supported" lists to allow custom codes, and added same locale comparison code to resource remaps (need to be tested).

Added custom locale selection dialog for the `Locale remap editor`.

With language and country lists by default:
<img width="1319" alt="loc_ed" src="https://user-images.githubusercontent.com/7645683/148050471-7a3d76c8-6444-4e2e-bd3b-6c0caf6bee5e.png">

And fully editable additional code components as an option:
<img width="1319" alt="loc_ed_adv" src="https://user-images.githubusercontent.com/7645683/148050481-e389670b-6d48-4eeb-8748-13d2ff4441dc.png">

Locale filters are moved from the locale editor to the selection dialog.

Also added property hint for locale, which is using the same dialog as an editor:
<img width="448" alt="prop_edit" src="https://user-images.githubusercontent.com/7645683/148048783-748d8f24-44c1-456a-9a8a-57320b85f5d7.png">

akien-mga commented on editor/editor_locale_dialog.cpp:
```diff
@@ -0,0 +1,551 @@
+			vb_lang_list->set_h_size_flags(Control::SIZE_EXPAND_FILL);
+			{
+				Label *lang_lbl = memnew(Label);
+				lang_lbl->set_text(TTR("Langauge:"));
```
Typo.

akien-mga commented:
Thanks!
